src/Compilers/Core/Portable/SourceGenerator.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    public interface ISourceGenerator
+    {
+        void Execute(SourceGeneratorContext context);
+    }
+
+    public struct SourceGeneratorContext
+    {
+        public Compilation Compilation { get; }
+
+        // TODO: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
+
+        public void AddSource(string fileNameHint, SourceText sourceText) { throw new NotImplementedException(); }
```

Comment:
We assume that some generators will want to generate more than one "file", for example in a 1:1 mapping for additional files. The hint here is intended to be used for a few reasons:

1. If the generated files are emitted to disk, having some ability to put some distinguishing text might be useful. For example, if you have two .resx files, generating the files with simply names of ResxGeneratedFile1.cs and ResxGeneratedFile2.cs wouldn't be terribly useful -- you'd want it to be something like ResxGeneratedFile-Strings.cs and ResxGeneratedFile-Icons.cs if you had two .resx files named "Strings" and "Icons" respectively.
2. The IDE needs some concept of a "stable" identifier. Source generators create a couple of fun problems for the IDE: users will want to be able to set breakpoints in a generated file, for example. If a source generator outputs multiple files we need to know which is which so we can know which file the breakpoints go with. A source generator of course is allowed to stop emitting a file if needs change (if you delete a .resx, then the generated file associated with it will also go away!), but this gives us some control here.

This was called "hint" in that the compiler is implicitly allowed to control the filename in however it ultimately needs, and if two source generators give the same "hint" it can still distinguish them with any sort of prefix/suffix as necessary.